### ISSUE TITLE
Rounding behavior changes between Ruby 2.7.2 and Ruby 3.0

### ISSUE THREAD

**Issue by egiurleo** [ruby/bigdecimal#185]
Hello! I am trying to upgrade my application to Ruby 3.0, and some of my tests are failing due to a change in behavior in the `BigDecimal#round` method between Ruby 2.7 and Ruby 3.0. Numbers that were previously rounded down are consistently rounded up in Ruby 3.0.

Here's an example of the change I'm experiencing:

```
➜ irb (Ruby 2.7.2)
require 'bigdecimal/util'
=> true
num = 98.00000000000001.to_d
=> 0.98e2
num.round(0, 1)
=> 0.98e2
```

```
➜ irb (Ruby 3.0.0)
require 'bigdecimal/util'
=> true
num = 98.00000000000001.to_d
=> 0.9800000000000001e2
num.round(0, 1)
=> 0.99e2
```

I'd love more context on this change -- was it intentional? And if so, is there something I can do in Ruby 3 to restore the previous behavior (maybe change the rounding mode?)

Thanks for your time and your hard work on this gem! :)

**Comment by unrooty** [comment]
@egiurleo did you figured out how to solve this?

**Comment by paracycle** [comment]
Running a bisect points to this commit as the source of the change: https://github.com/ruby/bigdecimal/commit/aa536cd4b585ff6045cf91f396bc3eef993b91f2 which points to #70 as the issue it is fixing. There is a long thread of conversation on that issue about the change in behaviour introduced by that commit. You might want to chime in on the thread there.

**Comment by clinejj** [comment]
I'm seeing a similar problem with Ruby 3.0.1:

```
irb (Ruby 2.7.3)
require 'bigdecimal/util'
=> true
num = 0.09
=> 0.09
num.to_d
=> 0.9e-1
```

```
irb (Ruby 3.0.1)
require 'bigdecimal/util'
=> true
num = 0.009
=> 0.009
num.to_d
=> 0.8999999999999999e-2
```

This can be resolved by manually passing in the precision that was specified before https://github.com/ruby/bigdecimal/commit/aa536cd4b585ff6045cf91f396bc3eef993b91f2 as @paracycle mentions:

```
irb (Ruby 3.0.1)
require 'bigdecimal/util'
=> true
0.009.to_d(Float::DIG)
=> 0.9e-2
0.009.to_d(Float::DIG+1)
=> 0.8999999999999999e-2
```

This is also breaking some tests that we had and is a pretty unexpected change for the default behavior.

**Comment by mikeyduece** [comment]
I'm also encountering an issue when changing from 1.3.5 to 3.0.2 for bigdecimal version. 
```ruby
1.3.5
0.07.to == 0.7e-1

3.0.2
0.07.to_d == 0.7000000000000001e-1
```

Im having a hard time trying to find the change that would cause this change

**Comment by kamipo** [comment]
Is this behavior considered a bug (actually it will be fixed by #180) like #192 is tagged as a bug?

In other words, should we test the `to_d` behavior change in third party libraries as an intentional behavior change in bigdecimal?

**Comment by soberstadt** [comment]
I would consider this a bug as it breaks existing behavior. What would be required to flag this issue as a bug?

**Comment by arthurwozniak** [comment]
Hi, any other suggestions with this? We ended up with patching Float class to act as it used to.

**Comment by mrkn** [comment]
The difference of the result of `98.00000000000001.to_d.round(0, 1)` comes from the change of `Float#to_d` introduced in bigdecimal 2.0.3.

```
$ RBENV_VERSION=2.7 ruby --disable=gems -rbigdecimal/util -ve "p BigDecimal::VERSION, 98.00000000000001.to_d"
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-linux]
"2.0.0"
0.98e2

$ RBENV_VERSION=2.7 ruby -rbigdecimal/util -ve "p BigDecimal::VERSION, 98.00000000000001.to_d"
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-linux]
"2.0.3"
0.9800000000000001e2
```

As you can see in the above, `98.00000000000001.to_d` in bigdecimal 2.0.0 doesn't preserve the precision of the original Float value.  From this buggy behavior, `98.00000000000001.to_d.round(0, 1)` is same as `98.to_d.round(0, 1)` in bigdecimal 2.0.0.

You can see the rounding behavior is not changed between bigdecimal 2.0.0 and the latest one by below.

```
$ RBENV_VERSION=2.7 ruby --disable=gems -rbigdecimal/util -ve "p BigDecimal::VERSION, BigDecimal('98.00000000000001').round(0, 1)"
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-linux]
"2.0.0"
0.99e2
$ RBENV_VERSION=3.0 ruby --disable=gems -rbigdecimal/util -ve "p BigDecimal::VERSION, BigDecimal('98.00000000000001').round(0, 1)"
ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [x86_64-linux]
"3.0.0"
0.99e2
```

**Comment by mrkn** [comment]
@kamipo 

> In other words, should we test the to_d behavior change in third party libraries as an intentional behavior change in bigdecimal?

`to_d` had a bug previously.  The behavior change in bigdecimal 2.0.3 is not an intentional behavior change, but just a bug fix.

**Comment by technicalpickles** [comment]
If this is considered a bug in a bugfix, would it be reasonable to port the fix back to the 3.0 line, so Ruby 3 can include it? When upgrading from Ruby < 3 to the latest, it's pretty common to jump one version at a time, ie to 3.0 before going to 3.1.

**Comment by mrkn** [comment]
@technicalpickles BigDecimal is one of the standard libraries, but it is also delivered as a gem library.  You can use the latest version of bigdecimal.gem by specifying the version constraint of bigdecimal in Gemfile even if you use Ruby 3.0.x.  This may be the easiest way, I think.

In another way, you can request to back-port the latest bigdecimal version into the maintenance branch of Ruby 3.0.x.  In this way, you need to create a back-port request ticket with the concrete reason in https://bugs.ruby-lang.org/.
